model/task_lifecycle.go

```diff
@@ -992,36 +992,64 @@ func RestartFailedTasks(opts RestartOptions) (RestartResults, error) {
 	if opts.IncludeSetupFailed {
 		failureTypes = append(failureTypes, evergreen.CommandTypeSetup)
 	}
-	tasksToRestart, err := task.Find(task.ByTimeStartedAndFailed(opts.StartTime, opts.EndTime, failureTypes))
+	tasksToRestart, err := task.FindWithDisplayTasks(task.ByTimeStartedAndFailed(opts.StartTime, opts.EndTime, failureTypes))
 	if err != nil {
 		return results, err
 	}
 
+	type taskGroupAndBuild struct {
+		Build     string
+		TaskGroup string
+	}
+	// only need to check one task per task group / build combination, and once per display task
+	taskGroupsToCheck := map[taskGroupAndBuild]string{}
+	displayTasksToCheck := map[string]bool{}
+	idsToRestart := []string{}
+	for _, t := range tasksToRestart {
+		if t.IsPartOfDisplay() {
+			displayTasksToCheck[t.DisplayTask.Id] = true
+		} else if t.DisplayOnly {
+			displayTasksToCheck[t.Id] = true
+		} else if t.IsPartOfSingleHostTaskGroup() {
+			taskGroupsToCheck[taskGroupAndBuild{
```

Comment:
I don't understand this part. Since tg + build doesn't uniquely identify anything, won't this constantly overwrite the value and end up restarting 1 random task in the group?

Reply:
Yeah--that's all we want to do, because the restart logic will restart the entire task group, so we don't want to do that n number of times.

Reply:
https://github.com/evergreen-ci/evergreen/blob/ba70a8a127408d5538037f3a0816124791aca9b4/model/task_lifecycle.go#L245